src/bregman.jl

```diff
@@ -25,68 +28,101 @@ Options structure for the bregman iteration algorithm
 
 """
-    bregman(A, TD, x, b, options)
+    bregman(A, x, b; options)
 
 Linearized bregman iteration for the system
 
 ``\\frac{1}{2} ||TD \\ x||_2^2 + λ ||TD \\ x||_1  \\ \\ \\ s.t Ax = b``
 
 For example, for sparsity promoting denoising (i.e LSRTM)
 
-# Arguments
+# Required arguments
 
-- `TD`: curvelet transform
-- `A`: Forward operator (J or preconditioned J for LSRTM)
-- `b`: observed data
+- `A`: Forward operator (e.g. J or preconditioned J for LSRTM)
 - `x`: Initial guess
+- `b`: observed data
+
+# Non-required arguments
+
+- `options`: bregman options, default is bregman_options()
 """
-function bregman(A, TD, x::Array{T}, b, options) where {T}
+
+function bregman(A, x::AbstractArray{T}, b, options=bregman_options()) where {T}
     # residual function wrapper
     function obj(x)
         d = A*x
         fun = .5*norm(d - b)^2
         grad = A'*(d - b)
         return fun, grad
     end
-    
-    return bregman(obj, x, options, TD)
+    return bregman(obj, x, options)
+end
+
+function bregman(A, TD, x::AbstractArray{T}, b, options=bregman_options()) where {T}
+    options.TD = TD
+    return bregman(A, x, b, options)
 end
 
 """
-    bregman(fun, TD, x, b, options)
+    bregman(funobj, x; options)
 
 Linearized bregman iteration for the system
 
 ``\\frac{1}{2} ||TD \\ x||_2^2 + λ ||TD \\ x||_1  \\ \\ \\ s.t Ax = b``
 
-For example, for sparsity promoting denoising (i.e LSRTM)
-
-# Arguments
+# Required arguments
 
-- `TD`: curvelet transform
-- `fun`: residual function, return the tuple (``f = \\frac{1}{2}||Ax - b||_2``, ``g = A^T(Ax - b)``)
-- `b`: observed data
+- `funobj`: a function that calculates the objective value (`0.5 * norm(Ax-b)^2`) and the gradient (`A'(Ax-b)`)
 - `x`: Initial guess
 
+# Non-required arguments
+
+- `options`: bregman options, default is bregman_options()
+- `TD`: sparsifying transform (e.g. curvelet), default is nothing (i.e. identity)
+- `λfunc`: a function to calculate the threshold in the 1st iteration, default is nothing
+- `λ`: a pre-determined threshold, will only be used if `λfunc` is not defined, default is nothing
+- `perc`: a percentage to calculate the threshold by quantile of the dual variable in 1st iteration, will only be used if neither `λfunc` nor `λ` are defined, default is .95
 """
-function bregman(funobj::Function, x::AbstractArray{T}, options::BregmanParams, TD=nothing) where {T}
+
+function bregman(funobj::Function, x::AbstractArray{T}, options=bregman_options()) where {T}
+    # set up how to calculate threshold in the first iteration
+    if isnothing(options.λfunc)
+        if ~isnothing(options.λ) 
+            λfunc = z->options.λ
+        else
+            λfunc = z->quantile(abs.(z), options.quantile)
+        end
+    end
+    return bregman(funobj, x, options, λfunc)
+end
+
+function bregman(funobj::Function, TD, x::AbstractArray{T}, options=bregman_options()) where {T}
+    options.TD = TD
```

Comment:
Mutating structure and objects is in general very bad practice coding-wise, so this needs a proper deprecation warning.